File: baselibrary/src/main/java/com/appbaselib/app/ActivityLifecycle.java
package com.appbaselib.app;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;
import android.support.v4.app.FragmentManager;

import java.util.List;

/*
 * Description: Activity生命周期注入
 * Created by Sum41forever on 2017/8/28
 * 修改备注:
 */
public class ActivityLifecycle implements Application.ActivityLifecycleCallbacks {

    private AppManager mAppManager;
    private Application mApplication;

    private List<FragmentManager.FragmentLifecycleCallbacks> mFragmentLifecycles;

    public ActivityLifecycle (Application application) {
        mAppManager = AppManager.getInstance();
        this.mApplication = application;
        mAppManager.setApplication(mApplication);
    }

    @Override
    public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
        //如果intent包含了此字段,并且为true说明不加入到list
        // 默认为false,如果不需要管理(比如不需要在退出所有activity(killAll)时，退出此activity就在intent加此字段为true)
        boolean isNotAdd = false;
        if (activity.getIntent() != null) {
            isNotAdd = activity.getIntent().getBooleanExtra(AppManager.IS_NOT_ADD_ACTIVITY_LIST, false);
        }
        if (!isNotAdd) {
            mAppManager.addActivity(activity);
        }

    }

    @Override
    public void onActivityStarted(Activity activity) {

    }

    @Override
    public void onActivityResumed(Activity activity) {
        mAppManager.setCurrentActivity(activity);
    }

    @Override
    public void onActivityPaused(Activity activity) {
        if (mAppManager.getCurrentActivity() == activity) {
            mAppManager.setCurrentActivity(null);
        }
    }

    @Override
    public void onActivityStopped(Activity activity) {

    }

    @Override
    public void onActivitySaveInstanceState(Activity activity, Bundle outState) {

    }

    @Override
    public void onActivityDestroyed(Activity activity) {
        mAppManager.removeActivity(activity);

    }

}

File: baselibrary/src/main/java/com/appbaselib/app/ActivityLifecycle.kt
package com.appbaselib.app

import android.app.Activity
import android.app.Application
import android.os.Bundle
import android.support.v4.app.FragmentManager

/*
 */
class ActivityLifecycle(private val mApplication: Application) : Application.ActivityLifecycleCallbacks {

    private val mAppManager: AppManager

    private val mFragmentLifecycles: List<FragmentManager.FragmentLifecycleCallbacks>? = null

    init {
        mAppManager = AppManager.getInstance()
        mAppManager.setApplication(mApplication)
    }

    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
        //如果intent包含了此字段,并且为true说明不加入到list
        // 默认为false,如果不需要管理(比如不需要在退出所有activity(killAll)时，退出此activity就在intent加此字段为true)
        var isNotAdd = false
        if (activity.intent != null) {
            isNotAdd = activity.intent.getBooleanExtra(AppManager.IS_NOT_ADD_ACTIVITY_LIST, false)
        }
        if (!isNotAdd) {
            mAppManager.addActivity(activity)
        }

    }

    override fun onActivityStarted(activity: Activity) {

    }

    override fun onActivityResumed(activity: Activity) {
        mAppManager.currentActivity = activity
    }

    override fun onActivityPaused(activity: Activity) {
        if (mAppManager.currentActivity === activity) {
            mAppManager.currentActivity = null
        }
    }

    override fun onActivityStopped(activity: Activity) {

    }

    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {

    }

    override fun onActivityDestroyed(activity: Activity) {
        mAppManager.removeActivity(activity)

    }

}
